DICOMPlugins/DICOMM3DPlugin.py

```diff
@@ -0,0 +1,641 @@
+# DICOM M3D objects can be for example .STL files encoded into a DICOM M3D object.
+#
+class DICOMM3DPluginClass(DICOMPluginBase):
+
+  def __init__(self):
+    super(DICOMM3DPluginClass,self).__init__()
+    self.loadType = "DICOMM3D"
+
+  def examineFiles(self,files):
+    """ Returns a list of DICOMLoadable instances
+    corresponding to ways of interpreting the
+    files parameter.
+    """
+    loadables = []
+
+    for candidateFile in files:
+
+      uid = slicer.dicomDatabase.fileValue(candidateFile, self.tags['instanceUID'])
+      if uid == '':
+        return []
+
+      desc = slicer.dicomDatabase.fileValue(candidateFile, self.tags['seriesDescription'])
+      if desc == '':
+        desc = "Unknown"
+
+      #read EncapsulatedDocument for candidateFile
+      docFile = self.getEncapsulatedDocument(candidateFile)
+
+      #read EncapsulatedDocumentLength for candidateFile
+      docLengthFile = self.getEncapsulatedDocumentLength(candidateFile)
+
+      #read modality type to flag M3D object.
+      isDicomM3D = (slicer.dicomDatabase.fileValue(candidateFile, self.tags['modality']) == 'M3D')
+
+      if isDicomM3D:
+        loadable = DICOMLoadable()
+        loadable.files = [candidateFile]
+        loadable.name = desc
+        loadable.tooltip = loadable.name + ' - as a DICOM M3D object'
+        loadable.selected = True
+        loadable.confidence = 0.95
+        loadable.uid = uid
+        loadable.doc = docFile
+        loadable.docLength = docLengthFile
+        self.addReferences(loadable)
+        loadables.append(loadable)
+
+        logging.debug('DICOM M3D modality found')
+
+    return loadables
+
+  def getFrameOfReferenceUID(self, candidateFile):
+    """Returns the frame of referenceUID for the given loadable"""
+    dcm = pydicom.read_file(candidateFile)
+    if hasattr(dcm, "FrameOfReferenceUID"):
+      return dcm.FrameOfReferenceUID
+    else:
+      return 'Unnamed FrameOfReferenceUID'
+  
+  def referencedSeriesName(self,loadable):
+    """Returns the default series name for the given loadable"""
+    referencedName = "Unnamed Reference"
+    if hasattr(loadable, "referencedSeriesUID"):
+      referencedName = self.defaultSeriesNodeName(loadable.referencedSeriesUID)
+    return referencedName
+
+  def getEncapsulatedDocument(self, candidateFile):
+    dcm = pydicom.read_file(candidateFile)
+    if hasattr(dcm, "EncapsulatedDocument"):
+      return dcm.EncapsulatedDocument
+    else:
+      return b''
+    
+  def getEncapsulatedDocumentLength(self, candidateFile):
+    dcm = pydicom.read_file(candidateFile)
+    if hasattr(dcm, "EncapsulatedDocumentLength"):
+      return dcm.EncapsulatedDocumentLength
+    else:
+      return 0
+
+  def load(self,loadable):
+    """ Load the DICOM M3D object
+    """
+    logging.debug('DICOM M3D load()')
+    try:
+      uid = loadable.uid
+      logging.debug('in load(): uid = '+uid)
+    except AttributeError:
+      return False
```

Comment:
It's better style (shorter, more efficient) to use hasattr rather than a try/except.